src/upgrade.cpp

```diff
@@ -0,0 +1,540 @@
+
+                        return false;
+                    }
+
+                    boost::filesystem::path ExtractFileString = ExtractPath / ZipStat.name;
+
+                    FILE* ExtractFile = fsbridge::fopen(ExtractFileString, "wb");
+
+                    if (!ExtractFile)
+                    {
+                        ExtractStatus.SnapshotExtractFailed = true;
+
+                        LogPrintf("Snapshot (ExtractSnapshot): Error opening file %s on filesystem", ZipStat.name);
+
+                        return false;
+                    }
+
+                    sum = 0;
+
+                    while ((uint64_t)sum != ZipStat.size)
+                    {
+                        boost::this_thread::interruption_point();
+
+                        len = zip_fread(ZipFile, Buf, 1024*1024);
+
+                        if (len < 0)
+                        {
+                            ExtractStatus.SnapshotExtractFailed = true;
+
+                            LogPrintf("Snapshot (ExtractSnapshot): Failed to read zip buffer");
+
+                            return false;
+                        }
+
+                        fwrite(Buf, 1, (uint64_t)len, ExtractFile);
+
+                        sum += len;
+                        currentuncompressedsize += len;
+
+                        // Update Progress every 1 second
+                        if (GetAdjustedTime() > lastupdated)
+                        {
+                            lastupdated = GetAdjustedTime();
+
+                            ExtractStatus.SnapshotExtractProgress = ((currentuncompressedsize / (double)totaluncompressedsize) * 100);
+                        }
+                    }
+
+                    fclose(ExtractFile);
+                    zip_fclose(ZipFile);
+                }
+            }
+        }
+
+        if (zip_close(ZipArchive) == -1)
+        {
+            ExtractStatus.SnapshotExtractFailed = true;
+
+            LogPrintf("Sanpshot (ExtractSnapshot): Failed to close snapshot.zip");
+
+            return false;
+        }
+
+        ExtractStatus.SnapshotExtractComplete = true;
+    }
+
+    catch (boost::thread_interrupted&)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+void Upgrade::DeleteSnapshot()
+{
+    // File is out of scope now check if it exists and if so delete it.
+    // This covers partial downloaded files or a http response downloaded into file.
+    std::string snapshotfile = GetArg("-snapshoturl", "https://download.gridcoin.us/download/downloadstake/signed/snapshot.zip");
+
+    size_t pos = snapshotfile.find_last_of("/");
+
+    snapshotfile = snapshotfile.substr(pos + 1, (snapshotfile.length() - pos - 1));
+
+    try
+    {
+        boost::filesystem::path snapshotpath = GetDataDir() / snapshotfile;
+
+        if (boost::filesystem::exists(snapshotpath))
+            if (boost::filesystem::is_regular_file(snapshotpath))
+                boost::filesystem::remove(snapshotpath);
```

Comment:
On Windows, this fails to remove the snapshot file. It looks like another issue with releasing the file handle: 

> Snapshot Downloader: Exception occured while attempting to delete snapshot (The process cannot access the file because it is being used by another process)

I don't see a place where the code isn't closing the snapshot.zip file. Do you have an idea?

Reply:
sha256sum verification function i left a file handle open i believe. i added `fclose(file);` there so please test that.

Reply:
The fix works for deleting snapshot.zip on Windows 👍 

One last issue: `snapshotfile` here should always be "snapshot.zip" because `Http::DownloadSnapshot()` always saves to that file, so we don't need the filename from the configuration. In fact, this method won't delete the snapshot.zip file when `-snapshoturl` contains a custom URL because that file won't exist.